Ported twisted.python._release to Python 3

rodrigc opened the pull request:
https://twistedmatrix.com/trac/ticket/8566

rodrigc committed:
Create a list around project.children()

On Python 3, project.children() returns a view, which we
cannot taken the length of without explicitly
creating a list.

rodrigc committed:
Use sorted() because a dict does not have a sort() method in Python 3

rodrigc committed:
The new template does not write out the project URL to index.html

However, it writes it out to ${PACKAGENAME}.html.
Fix the test to reflect this.

rodrigc committed:
Use bytes when writing to file object.

Decode from bytes when reading from file object.

rodrigc committed:
Allow test_release to run on Python 3

rodrigc committed:
Add topfile

rodrigc committed:
Look in index.html

rodrigc committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

glyph commented on src/twisted/python/_release.py:
```diff
@@ -39,14 +40,50 @@
     kwargs['stderr'] = STDOUT
-    return check_output(args, **kwargs)
+    output = check_output(args, **kwargs)
+    if not isinstance(output, str):
```
The care that is being taken with the output type here should be reflected in the docstring.  What type does `runCommand` return?

rodrigc replied:
runCommand returns the return value of subprocess.check_output(): https://docs.python.org/3/library/subprocess.html#subprocess.check_output

glyph replied:
Well, yes, but what I mean is:

1. this should be documented in its docstring as an `@rtype`
2. its type, specifically, should be documented.  is it supposed to be `L{bytes}` or `L{unicode}`?

glyph commented on src/twisted/python/_release.py:
```diff
@@ -158,7 +195,7 @@ def exportTo(fromDir, exportDir):
         runCommand(["git", "-C", fromDir.path,
                     "checkout-index", "--all", "--force",
-                    # prefix has to end up with a "/" so that files get copied
+                    # Prefix has to end up with a "/" so that files get copied
```
I'm assuming that this unrelated change is to quiet down twistedchecker :).  If you want to make a code-identifier the subject of a sentence, you can fake out twistedchecker by using punctuation at the start of the sentence.  Often this means you can use `` `backticks` `` around the identifier, but in this case, the more precise / correct fix is to just say `--prefix has to end up with ...`, which twistedchecker (by design) won't complain about.

glyph commented on src/twisted/python/_release.py:
```diff
@@ -211,6 +248,7 @@ def __repr__(self):
 
     def getVersion(self):
         """
+
```
This seems wrong to me - docstrings shouldn't have to start with a blank line. Is twistedchecker complaining about it?  If so perhaps we should file a bug; "`@return` is the only thing in a docstring" seems like an OK thing that pydoctor could be corrected to understand (making the `@return` the summary, etc).

glyph commented on src/twisted/python/_release.py:
```diff
@@ -482,13 +532,19 @@ def _writeMisc(self, fileObj, header, tickets):
 
-        fileObj.write(header + '\n' + '-' * len(header) + '\n')
+        line = header + '\n' + '-' * len(header) + '\n'
+        if not isinstance(line, bytes):
```
`line` is a variable that you just synthesized; you should know its type.  Rather than `isinstance`-checking it after creating it (a scenario which I believe can only occur on python2 where implicit encoding is allowed), you should use something like [`_matchingString`](https://twistedmatrix.com/documents/16.6.0/api/twisted.python.compat.html#_matchingString) for the `header`, and explicit `b` prefixes.

glyph commented on src/twisted/python/_release.py:
```diff
@@ -39,14 +40,50 @@
+
+
+
+def setStrContent(fileObj, content, encoding="utf-8"):
```
Naming suggestion: I like using the word `lenient` in functions like this, because it explicitly makes it clear that it's accepting of multiple types.

glyph commented on src/twisted/python/test/test_release.py:
```diff
@@ -379,43 +390,45 @@ def test_build(self):
             "def _bar():\n"
             "    '%s'" % (docstring, privateDocstring))
-
+        setStrContent(inputPath.child("__init__.py"), inputPathContent)
```
This binds another name in a very long function, which means one more thing for a reader to scan up and down to see if it's used.

My preference would be to do something like

```python
        setStrContent(
              inputPath.child("__init__.py"),
              "def foo():\n"		              "def foo():\n"
              "    '%s'\n"		              "    '%s'\n"
              "def _bar():\n"		              "def _bar():\n"
              "    '%s'" % (docstring, privateDocstring)
        )
```
to keep it an expression and avoid the subsequently unused `inputPathContent` name.

glyph commented on src/twisted/python/test/test_release.py:
```diff
@@ -734,8 +755,11 @@ def test_writeHeader(self):
         output = StringIO()
         self.builder._writeHeader(output, "Super Awesometastic 32.16")
+        content = output.getvalue()
+        if not isinstance(content, str):
```
`isinstance` checks in tests are the worst, since they morph the tests to test different things on different versions of Python, which is easy to screw up in a way where you're testing the wrong thing.  I think it's right here, but I'm not even confident in my ability to always catch that as a reviewer.  You're always using a `BytesIO` here, so you know that it's going to be `bytes`; perhaps just a `b` prefix on the test value wold do what you want?

glyph reviewed (commented):
Thanks for porting this!

The major issue is the failing tests everywhere, and that obviously needs to be fixed.  But in the interests of keeping you unblocked I made sure to do a full review as well.

If the tests were passing, this would all be optional feedback, since the changes are entirely in non-public code and therefore interface compatibility is not a concern, but the failures do necessitate a re-review.

So please address the test failures, and then take any of the comments here into consideration as you see fit, and then resubmit for review.

Thanks again!  Looking forward to deleting the last entry from that list of modules :).

rodrigc committed:
Rename to lenientSetStrContent()

rodrigc committed:
Fix comment

rodrigc committed:
Fix docstring

rodrigc committed:
Remove debugging sleep

rodrigc committed:
Explicitly use io.BytesIO

rodrigc committed:
When overriding sys.stdout, use NativeStringIO.

sys.stdout is str on Python 3 and bytes on Python 2

rodrigc committed:
Use twisted.python.compat._matchingString

rodrigc committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

rodrigc committed:
Shorten lines

glyph committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

glyph commented on src/twisted/python/_release.py:
```diff
@@ -39,14 +40,50 @@
+
+def getStrContent(fileObj, encoding="utf-8"):
+    """
+    Get the content from a L{FilePath} and return it as L{str}.
```
Since this is designed specifically to return it as a `str` as opposed to `bytes` or `unicode`, it should use the disambiguating prefix "native".  Older code that is confused about text/bytes often documents types as `str` and that's sometimes vague when it comes to how other cross-version code should expect to consume it.

glyph commented:
The coverage issues look like... some of the _tests_ aren't even running?

rodrigc committed:
Update docstring

rodrigc committed:
Remove debugging sleep

rodrigc committed:
Rename getStrContent to getNativeStrContent

rodrigc committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

rodrigc committed:
Merge branch '8566-rodrigc-release-py3-2' of https://github.com/twisted/twisted into 8566-rodrigc-release-py3-2

rodrigc committed:
getNativeStrContent

rodrigc committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

rodrigc committed:
Fix

rodrigc committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

rodrigc committed:
Fix some pylint warnings

rodrigc committed:
Disable some pylint errors raising false positives.

 twistedchecker falsely complains when # is inside a string, thinking it is a comment.

rodrigc committed:
Merge branch 'trunk' into 8566-rodrigc-release-py3-2

rodrigc committed:
Need to disable one more pylint false positive.

rodrigc committed:
Fix typo

rodrigc commented:
Superseded by https://github.com/twisted/twisted/pull/713
